SE-3310 Ensure prod instances email urgent on monitoring fail and alerts auto close once resolved

samuelallan72 opened the pull request:
Otherwise we only get paged on a provision failure; ping monitoring
previously would only email the settings.ADMINS emails.

**Test instructions**:

- register a new account, choose instance settings, and confirm your email
- view the settings of the new instance
- verify that additional_monitoring_emails has the PROD_APPSERVER_FAIL_EMAILS added
- verify that the corresponding newrelic alert policy for the new instance has ADMINS _and_ PROD_APPSERVER_FAIL_EMAILS set in the notification channels
- do something to take the instance offline (change the dns records, stop the lms service, etc.)
- verify that newrelic alerts open
- put the instance back online
- verify the newrelic alerts/incidents auto close after a few minutes.

**Reviewers**:

- [x] @giovannicimolin 
- [ ] @lgp171188

samuelallan72 committed:
Ensure prod instances email urgent on monitoring fail

Otherwise we only get paged on a provision failure; ping monitoring
previously would only email the settings.ADMINS emails.

lgp171188 commented on registration/provision.py:
```diff
@@ -93,6 +93,7 @@ def _provision_instance(sender, **kwargs):
 
         if settings.PROD_APPSERVER_FAIL_EMAILS:
             application.instance.provisioning_failure_notification_emails = settings.PROD_APPSERVER_FAIL_EMAILS
+            application.instance.additional_monitoring_emails = settings.PROD_APPSERVER_FAIL_EMAILS
```
@swalladge, afaik, the monitoring SLA is applicable only for paying customers - P&T and institutional clients and that is why we add `urgent@` manually to those instances.

giovannicimolin replied:
@lgp171188 This is the `production_instance_factory`, so it's fine. If a developer ends up spawning an appserver that fails, it's his responsibility of cleaning it up/addressing the page. 
I'd prefer keep this explicit mechanism instead of having to manually set the monitoring.

giovannicimolin replied:
Btw @swalladge, take a look at this: https://github.com/open-craft/opencraft/pull/489 (your previous PR here).
Do we really need this change here? I assumed we needed to set the emails manually, but for P&T, they are already added, so no problem.

swalladge replied:
@giovannicimolin according to the original requirements, yes we did need this change.

> Doing this because we only want to auto set PROD_APPSERVER_FAIL_EMAILS for end user instances.

You can check SE-1832, but I believe the reason to move things to here, is that now they will be autoset up for new instances that are set up through the betatest registration process (our standard documented process for setting up client and maintained instances).

swalladge replied:
Also, provisioning_failure_notification_emails != additional_monitoring_emails, so I'm not sure what you're asking. :thinking:

giovannicimolin replied:
Yup, I think I've mixed up those two.

samuelallan72 committed:
Use sum instead of single value for NR alerts

This should allow the alerts to be auto closed.

With the original single_value setup,
alerts would only be autoclosed if _every_ query returned at least 1
for a full 16 minutes.
This is impossible, because ping monitors ping in only 5 minute
intervals, and the range for a single query is 1 minute,
so there will be at most 2 successive queries returning at least 1.

By switching to sum, it has the same behaviour of opening an alert,
but the auto close condition becomes 'at least 1 in the last 16 minutes',
which is an easy condition to meet once the instance comes back online.

giovannicimolin reviewed (approved):
:+1:

- [x] I tested this:
1. Tested the change on a live monitor in Ocim production.
- [x] I read through the code: simple diff.
- [ ] ~~I checked for accessibility issues~~ NA
- [ ] ~~Includes documentation~~ NA

@swalladge Good to merge & test on production monitors.

swalladge commented:
Merging because the integration test failures are unrelated.
